Mathlib/Topology/UniformSpace/Completion.lean

```diff
@@ -0,0 +1,728 @@
+#align uniform_space.completion.dense_range_coe₃ UniformSpace.Completion.denseRange_coe₃
+
+@[elab_as_elim]
+theorem induction_on {p : Completion α → Prop} (a : Completion α) (hp : IsClosed { a | p a })
+    (ih : ∀ a : α, p a) : p a :=
+  isClosed_property denseRange_coe hp ih a
+#align uniform_space.completion.induction_on UniformSpace.Completion.induction_on
+
+@[elab_as_elim]
+theorem induction_on₂ {p : Completion α → Completion β → Prop} (a : Completion α) (b : Completion β)
+    (hp : IsClosed { x : Completion α × Completion β | p x.1 x.2 })
+    (ih : ∀ (a : α) (b : β), p a b) : p a b :=
+  have : ∀ x : Completion α × Completion β, p x.1 x.2 :=
+    isClosed_property denseRange_coe₂ hp fun ⟨a, b⟩ => ih a b
+  this (a, b)
+#align uniform_space.completion.induction_on₂ UniformSpace.Completion.induction_on₂
+
+@[elab_as_elim]
+theorem induction_on₃ {p : Completion α → Completion β → Completion γ → Prop} (a : Completion α)
+    (b : Completion β) (c : Completion γ)
+    (hp : IsClosed { x : Completion α × Completion β × Completion γ | p x.1 x.2.1 x.2.2 })
+    (ih : ∀ (a : α) (b : β) (c : γ), p a b c) : p a b c :=
+  have : ∀ x : Completion α × Completion β × Completion γ, p x.1 x.2.1 x.2.2 :=
+    isClosed_property denseRange_coe₃ hp fun ⟨a, b, c⟩ => ih a b c
+  this (a, b, c)
+#align uniform_space.completion.induction_on₃ UniformSpace.Completion.induction_on₃
+
+theorem ext {Y : Type _} [TopologicalSpace Y] [T2Space Y] {f g : Completion α → Y}
+    (hf : Continuous f) (hg : Continuous g) (h : ∀ a : α, f a = g a) : f = g :=
+  cPkg.funext hf hg h
+#align uniform_space.completion.ext UniformSpace.Completion.ext
+
+theorem ext' {Y : Type _} [TopologicalSpace Y] [T2Space Y] {f g : Completion α → Y}
+    (hf : Continuous f) (hg : Continuous g) (h : ∀ a : α, f a = g a) (a : Completion α) :
+    f a = g a :=
+  congr_fun (ext hf hg h) a
+#align uniform_space.completion.ext' UniformSpace.Completion.ext'
+
+section Extension
+
+variable {f : α → β}
+
+/-- "Extension" to the completion. It is defined for any map `f` but
+returns an arbitrary constant value if `f` is not uniformly continuous -/
+protected def extension (f : α → β) : Completion α → β :=
+  cPkg.extend f
+#align uniform_space.completion.extension UniformSpace.Completion.extension
+
+section CompleteSpace
+
+variable [CompleteSpace β]
+
+theorem uniformContinuous_extension : UniformContinuous (Completion.extension f) :=
+  cPkg.uniformContinuous_extend
+#align uniform_space.completion.uniform_continuous_extension UniformSpace.Completion.uniformContinuous_extension
+
+theorem continuous_extension : Continuous (Completion.extension f) :=
+  cPkg.continuous_extend
+#align uniform_space.completion.continuous_extension UniformSpace.Completion.continuous_extension
+
+end CompleteSpace
+
+/- porting note: removed `@[simp]` because this lemma doesn't even trigger on itself in Lean 3 or
+Lean 4 unless the user manually supplies the `hf` argument, so it is useless as a `simp` lemma. -/
```

Comment:
Did the linter complain? Usually the linter is smart enough to pass `hf` to simp. If `simp [hf]` makes the lemma trigger on itself, then the linter should not complain.

Reply:
Yes the linter complained for all three lemmas.